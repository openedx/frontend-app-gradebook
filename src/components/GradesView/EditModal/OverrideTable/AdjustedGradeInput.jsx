/* eslint-disable react/sort-comp, react/button-has-type */
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { Form } from '@edx/paragon';

import selectors from 'data/selectors';
import actions from 'data/actions';
import { getLocale, isRtl } from '@edx/frontend-platform/i18n';

/**
 * <AdjustedGradeInput />
 * Input control for adjusting the grade of a unit
 * displays an "/ ${possibleGrade} if there is one in the data model.
 */
export class AdjustedGradeInput extends React.Component {
  constructor(props) {
    super(props);
    this.onChange = this.onChange.bind(this);
  }

  onChange = ({ target }) => {
    let adjustedGradeValue;
    switch (true) {
      case target.value < 0:
        adjustedGradeValue = 0;
        break;
      case this.props.possibleGrade && target.value > this.props.possibleGrade:
        adjustedGradeValue = this.props.possibleGrade;
        break;
      default:
        adjustedGradeValue = target.value;
    }
    this.props.setModalState({ adjustedGradeValue });
  };

  render() {
    return (
      <span>
        <Form.Control
          type="number"
          name="adjustedGradeValue"
          min="0"
          max={this.props.possibleGrade ? this.props.possibleGrade : ''}
          value={this.props.value}
          onChange={this.onChange}
        />
        {this.props.possibleGrade && ` ${isRtl(getLocale()) ? '\\' : '/'} ${this.props.possibleGrade}`}
      </span>
    );
  }
}
AdjustedGradeInput.defaultProps = {
  possibleGrade: null,
};
AdjustedGradeInput.propTypes = {
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]).isRequired,
  possibleGrade: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  setModalState: PropTypes.func.isRequired,
};

export const mapStateToProps = (state) => ({
  possibleGrade: selectors.root.editModalPossibleGrade(state),
  value: selectors.app.modalState.adjustedGradeValue(state),
});

export const mapDispatchToProps = {
  setModalState: actions.app.setModalState,
};

export default connect(mapStateToProps, mapDispatchToProps)(AdjustedGradeInput);
