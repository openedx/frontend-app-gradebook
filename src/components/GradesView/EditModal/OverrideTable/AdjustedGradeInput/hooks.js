import { actions, selectors } from 'data/redux/hooks';
import { getLocalizedSlash } from 'i18n/utils';

const useAdjustedGradeInputData = () => {
  const possibleGrade = selectors.root.useEditModalPossibleGrade();
  const value = selectors.app.useModalData().adjustedGradeValue;
  const setModalState = actions.app.useSetModalState();
  const hintText = possibleGrade && ` ${getLocalizedSlash()} ${possibleGrade}`;

  const onChange = ({ target }) => {
    setModalState({ adjustedGradeValue: target.value });
  };

  return {
    value,
    onChange,
    hintText,
  };
};

export default useAdjustedGradeInputData;
