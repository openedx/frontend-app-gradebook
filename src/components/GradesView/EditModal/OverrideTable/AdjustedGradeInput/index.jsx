import React from 'react';

import { Form } from '@edx/paragon';

import useAdjustedGradeInputData from './hooks';

/**
 * <AdjustedGradeInput />
 * Input control for adjusting the grade of a unit
 * displays an "/ ${possibleGrade} if there is one in the data model.
 */
export const AdjustedGradeInput = () => {
  const {
    value,
    onChange,
    hintText,
  } = useAdjustedGradeInputData();
  return (
    <span>
      <Form.Control
        type="text"
        name="adjustedGradeValue"
        value={value}
        onChange={onChange}
      />
      {hintText}
    </span>
  );
};

AdjustedGradeInput.propTypes = {};

export default AdjustedGradeInput;
