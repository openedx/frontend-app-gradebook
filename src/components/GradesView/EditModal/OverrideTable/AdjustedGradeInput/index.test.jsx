import React from 'react';
import { shallow } from 'enzyme';

import { Form } from '@edx/paragon';

import useAdjustedGradeInputData from './hooks';
import AdjustedGradeInput from '.';

jest.mock('./hooks', () => jest.fn());

const hookProps = {
  hintText: 'some-hint-text',
  onChange: jest.fn().mockName('hook.onChange'),
  value: 'test-value',
};
useAdjustedGradeInputData.mockReturnValue(hookProps);

let el;
describe('AdjustedGradeInput component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    el = shallow(<AdjustedGradeInput />);
  });
  describe('behavior', () => {
    it('initializes hook data', () => {
      expect(useAdjustedGradeInputData).toHaveBeenCalled();
    });
  });
  describe('render', () => {
    test('snapshot', () => {
      expect(el).toMatchSnapshot();
      const control = el.find(Form.Control);
      expect(control.props().value).toEqual(hookProps.value);
      expect(control.props().onChange).toEqual(hookProps.onChange);
      expect(el.contains(hookProps.hintText)).toEqual(true);
    });
  });
});
