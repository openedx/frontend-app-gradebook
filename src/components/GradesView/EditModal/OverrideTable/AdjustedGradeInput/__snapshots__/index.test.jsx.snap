// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdjustedGradeInput component render snapshot 1`] = `
<span>
  <Form.Control
    name="adjustedGradeValue"
    onChange={[MockFunction hook.onChange]}
    type="text"
    value="test-value"
  />
  some-hint-text
</span>
`;
